### PR TITLE
If token validation fails, give the user the expected format.

### PR DESCRIPTION
**Pull request by pipejakob**
If a user specifies their own token to kubeadm, and it fails validation, the error they currently receive isn't the friendliest. This first change adds messaging for the expected token format, with more improvements to follow as part of #33930. It also adds some unit tests to document the behavior we have currently.

CC: @mikedanese, @krousey, @kubernetes/sig-cluster-lifecycle

Part of #33930.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35075)

### CODE REVIEW

File: cmd/kubeadm/app/util/tokens_test.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+)
+
+func TestUsingEmptyTokenFails(t *testing.T) {
+	// Simulates what happens when you omit --token on the CLI
+	s := newSecretsWithToken("")
+
+	ok, err := UseGivenTokenIfValid(s)
+	assert.False(t, ok)
+	assert.NoError(t, err)
```
Comment (krousey) [inline cmd/kubeadm/app/util/tokens_test.go]: Generally you want to check for errors first. If an error occurred, what does `ok` even mean?

Also, the assert library has some very basic and uninformative failure messages like "Unexpected error" and "Should be false". Please provide more context to these failures. Ideally, I shouldn't have to read the code to know the general nature of the failure.

My preference is to use standard if/else comparisons, call `testing.T`'s methods directly, and avoid assertion libraries because:

- They make it easy to omit useful messages
- They add little value over standard `if` checks.
- They often conflate comparison utilities with test failures. For example, this library has `assert.JSONEq`. It does not have `assert.JSONNotEq`. However if I had `func compareJSON(a, b string) bool`, the difference between testing equality and inequality is adding a ! in front of it. These libraries tend to end up with a cross product of comparisons and boolean operations.

File: cmd/kubeadm/app/util/tokens_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		"123456.1234567890.123456",
+	}
+
+	for _, token := range(invalidTokens) {
+		s := newSecretsWithToken(token)
+		ok, err := UseGivenTokenIfValid(s)
+
+		assert.False(t, ok, "expected invalid token to return ok = false: [%s]", token)
+		assert.Error(t, err, "expected invalid token to return an error: [%s]", token)
+		assert.Contains(t, err.Error(), "<6 characters>.<16 characters>", "expected better validation failure message")
```
Comment (krousey) [inline cmd/kubeadm/app/util/tokens_test.go]: I don't think we need a test to verify you typed what you think you typed.

However, if this method could return different types of errors, those tests would be useful. And generally, inspecting errors strings is a bad idea. They are meant for human consumption. If you need to programatically signal different types of failures through errors, there are better ways. http://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully does a good job of enumerating different approaches to having really useful errors.

**Review by krousey — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/util/tokens_test.go (hunk 2 above)
Comment (pipejakob) [inline cmd/kubeadm/app/util/tokens_test.go]: Yeah, this was meant to be an acceptance test. I'm not sure if the project is big on those (I'm new), but I have a habit of wanting to write tests that demonstrate exactly what behavior I've changed, and the only behavior changed in this PR was the human error message. I can rip the test out.

Comment (pires) [inline cmd/kubeadm/app/util/tokens_test.go]: We should have error types and not compare strings because that's hugely error prone.

File: cmd/kubeadm/app/util/tokens_test.go (hunk 1 above)
Comment (krousey) [inline cmd/kubeadm/app/util/tokens_test.go]: I want to preface this with this is just debating opinions here. I think we both agree the error message needs improving, and that's all I really want. What follows is probably a bit of a rant.

> I don't think I agree with the overall sentiment about assertion libraries. I can absolutely add better messaging for these assertions, and you're right to call that out, but if you have a few things to assert in a test, I think Go's if statements (with required curly braces) make it far less semantically compact to indicate their function to the reader of the code.

Being compact does not necessarily make it easier for the reader, especially when the `if/else` checks in the test would more closely resemble the code under test with all the 

```go
if err != nil {
	// handle error
}
```

error handling. And honestly it's only saving 2 lines of code. Moreover, the goal should be that the error message is descriptive enough that I shouldn't have to read the test code too closely to determine what went wrong.

> Especially functions like "assert.Equals" can remove a lot of the boilerplate code associated with not only checking equality, but printing a nicely-formatted message including the expected vs. got for the variables instead of needing to construct the format yourself by hand.

This is what `assert.False` prints by default:

```console
	Error Trace:	tokens_test.go:31
	Error:		Should be false
```

You can add a message to make it something like this:

```console
	Error Trace:	tokens_test.go:31
	Error:		Should be false
	Messages:	UseGivenTokenIfValid indicated an empty token is valid
```

That's very verbose and mostly useless. Compare that to

```go
if got := UseGivenTokenIfValid(s); got != want {
	test.Errorf("UseGivenTokenIfValid indicated an empty token is valid: returned true, expected false")
}
```

```console
	tokens_test.go:31: UseGivenTokenIfValid indicated an empty token is valid: returned true, expected false
```

Concise one-line output. As a side note: this also a standard format for printing errors that most editors understand by default and can easily navigate to the line the test failed on. (That format being a line starting with file_name:line_no)

And then there's the complexity these libraries tend towards. I think it's best summed up here: https://golang.org/doc/faq#testing_framework

> A related point is that testing frameworks tend to develop into mini-languages of their own, with conditionals and controls and printing mechanisms, but Go already has all those capabilities; why recreate them? We'd rather write tests in Go; it's one fewer language to learn and the approach keeps the tests straightforward and easy to understand.

> Since this personal preference can easily swing both ways, can we put a ruling into the Coding Conventions doc? If the project preference is already documented somewhere, then sorry for missing it.

I thought we already fell back to the go standards, but I see we only call out Effective Go. But overall, this is something I'm more than willing to debate, but I do not want to mandate it.


TL;DR make the messages better. Be aware of the different ways of doing these things and the various trade-offs.

File: cmd/kubeadm/app/util/tokens_test.go (hunk 2 above)
Comment (pipejakob) [inline cmd/kubeadm/app/util/tokens_test.go]: Yup, I will absolutely rip this test out.

File: cmd/kubeadm/app/util/tokens_test.go (hunk 1 above)
Comment (pipejakob) [inline cmd/kubeadm/app/util/tokens_test.go]: I'd love to discuss this more, but I think offline face-to-face might be more effective. Similar to coding style, this is the kind of thing that I think merits exactly one discussion per project, and then the consensus should just be documented as the project's standard. Otherwise, without a standard there will be no consistency in the codebase, and the same discussion will keep popping up and eating developer cycles whenever someone from camp A reviews code from someone in camp B and they have to rehash the same points. As much as I like unit test assertions, I far prefer consistency and putting the discussion to rest. :-)

I wonder if there's a happy medium, though, to create our own internal assertion helpers that address the points you mentioned. They can make failure descriptions required, have very concise output, and avoid the pitfalls of including anything like JSONEq, while still delivering more succinct checks than manual ifs in tests. Happy to discuss more.

For now, though, I'll ditch the assertion library and give vanilla Go testing a try. Thanks!

File: cmd/kubeadm/app/util/tokens_test.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+func TestUsingEmptyTokenFails(t *testing.T) {
+	// Simulates what happens when you omit --token on the CLI
+	s := newSecretsWithToken("")
+
+	given, err := UseGivenTokenIfValid(s)
+	if err != nil {
+		t.Errorf("UseGivenTokenIfValid returned an error when the token was omitted: %v", err)
+	}
+	if given {
+		t.Errorf("UseGivenTokenIfValid returned given = true when the token was omitted; expected false")
```
Comment (krousey) [inline cmd/kubeadm/app/util/tokens_test.go]: Minor nit. Since you're not formatting in this call, you can just use `t.Error`

**Review by krousey — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/util/tokens_test.go (hunk 3 above)
Comment (pipejakob) [inline cmd/kubeadm/app/util/tokens_test.go]: Nice catch. Dropped the f.

**Comment by jbeda** [comment]
Feel free to fix nit.  I'm cool with it as is.

**Comment by pipejakob** [comment]
@k8s-bot cvm gke e2e test this

**Comment by pipejakob** [comment]
k8s-bot gci gke e2e test this

**Comment by pipejakob** [comment]
@k8s-bot gci gke e2e test this
